clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2175,14 +2178,23 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
 
   void addParam(const FieldDecl *FD, QualType ArgTy,
                 SYCLIntegrationHeader::kernel_param_kind_t Kind) {
+    addParam(FD, ArgTy, Kind, offsetOf(FD, ArgTy));
+  }
+  void addParam(const FieldDecl *FD, QualType ArgTy,
+                SYCLIntegrationHeader::kernel_param_kind_t Kind,
+                uint64_t OffsetAdj) {
     uint64_t Size;
-    const ConstantArrayType *CAT =
-        SemaRef.getASTContext().getAsConstantArrayType(ArgTy);
-    if (CAT)
-      ArgTy = CAT->getElementType();
     Size = SemaRef.getASTContext().getTypeSizeInChars(ArgTy).getQuantity();
     Header.addParamDesc(Kind, static_cast<unsigned>(Size),
-                        static_cast<unsigned>(CurOffset + offsetOf(FD)));
+                        static_cast<unsigned>(CurOffset + OffsetAdj));
+  }
+
+  // Returns 'true' if the thing we're visiting (Based on the FD/QualType pair)
+  // is an element of an array.  This will determine whether we do
+  // MemberExprBases in some cases or not, AND determines how we initialize
+  // values.
+  bool IsArrayElement(const FieldDecl *FD, QualType Ty) const {
+    return !SemaRef.getASTContext().hasSameType(FD->getType(), Ty);
```

Comment:
Array is the only case where handler function receives type and field of other type, right?

Reply:
Yes, exactly.  It is the only case where the FieldDecl doesn't represent the thing that we're visiting.  I would like some day to come up with a better way to differentiate, but its not obvious to me how to do this without some sort of state being set in the visitor.

Reply:
For pointers, the address space qualifier can be different for FD and the type we generated in DeclCreator. I assume that doesn't matter here?

Reply:
Right, it shouldn't matter.  The isArrayElement check happens on the type that we get from the visitor.